primitives/src/erasure_code/reed_solomon_erasure.rs

```diff
@@ -0,0 +1,218 @@
+// Copyright (c) 2022 Espresso Systems (espressosys.com)
+// This file is part of the Jellyfish library.
+
+// You should have received a copy of the MIT License
+// along with the Jellyfish library. If not, see <https://mit-license.org/>.
+
+//! Module for Reed Solomon Erasure Code
+
+use crate::errors::PrimitivesError;
+use ark_ff::PrimeField;
+use ark_serialize::{CanonicalDeserialize, CanonicalSerialize};
+use ark_std::{borrow::Borrow, string::ToString, vec, vec::Vec};
+use core::marker::PhantomData;
+
+use super::ErasureCode;
+
+/// Very naive implementation of Reed Solomon erasure code.
+///  * `reconstruction_size`: and the minimum number of shards required for
+///    reconstruction
+///  * `num_shards`: the block (codeword) length
+/// The encoding of a message is the evaluation on (1..num_shards) of the
+/// polynomial whose coefficients are the message entries. Decoding is a naive
+/// Lagrange interpolation.
+pub struct ReedSolomonErasureCode<F: PrimeField> {
+    reconstruction_size: usize,
+    num_shards: usize,
+    phantom_f: PhantomData<F>,
+}
+
+/// Shards for Reed Solomon erasure code
+#[derive(Clone, Eq, Debug, CanonicalSerialize, CanonicalDeserialize, Derivative)]
+#[derivative(PartialEq, Hash(bound = "F: PrimeField"))]
+pub struct ReedSolomonErasureCodeShard<F: PrimeField> {
+    /// Index of shard shard
+    pub index: usize,
+    /// Value of this shard
+    pub values: Vec<F>,
+}
+
+impl<F: PrimeField> ReedSolomonErasureCodeShard<F> {
+    /// Create a new shard
+    pub fn new(index: usize, values: Vec<F>) -> Self {
+        Self { index, values }
+    }
+}
+
+impl<F: PrimeField> ErasureCode for ReedSolomonErasureCode<F> {
+    type Field = F;
+    type Shard = ReedSolomonErasureCodeShard<F>;
+
+    fn new(reconstruction_size: usize, num_shards: usize) -> Result<Self, PrimitivesError> {
+        if reconstruction_size > num_shards {
+            Err(PrimitivesError::ParameterError(
+                "Number of shards must be at least the message length.".to_string(),
+            ))
+        } else {
+            Ok(ReedSolomonErasureCode {
+                reconstruction_size,
+                num_shards,
+                phantom_f: PhantomData,
+            })
+        }
+    }
+
+    /// The encoding will split the data into chunks of length
+    /// `reconstruction_size`. And represent each chunk as a polynomial. The
+    /// codeword composes of evaluations of those polynomials on
+    /// (1..num_shards).
+    fn encode(&self, data: &[Self::Field]) -> Result<Vec<Self::Shard>, PrimitivesError> {
```

Comment:
Should we use a `Polynomial` trait here instead of just slice of `Field`? [polynomial trait](https://github.com/EspressoSystems/jellyfish/blob/31ca8d9b585d9de55c0bda7c3bc1795543a49c84/primitives/src/pcs/mod.rs#L38) ?